server/src-lib/Hasura/RQL/DDL/Schema/Function.hs

```diff
@@ -0,0 +1,196 @@
+  assertTableExists retTable $ "return type " <> retTable <<> " is not a valid table"
+
+  -- inpArgTyps <- mapM fetchTypNameFromOid inpArgTypIds
+  let funcArgs = Seq.fromList $ mkFunctionArgs inpArgTyps inpArgNames
+      dep = SchemaDependency (SOTable retTable) "table"
+  return $ FunctionInfo qf False funTy funcArgs retTable [dep]
+  where
+    RawFuncInfo hasVariadic funTy retSn retN retTyTyp retSet inpArgTyps inpArgNames = rawFuncInfo
+
+-- Build function info
+getFunctionInfo :: QualifiedFunction -> Q.TxE QErr FunctionInfo
+getFunctionInfo qf@(QualifiedObject sn fn) = do
+  -- fetch function details
+  funcData <- Q.listQE defaultTxErrorHandler [Q.sql|
+        SELECT
+          row_to_json (
+            (
+              SELECT
+                e
+              FROM
+                (
+                  SELECT
+                    has_variadic,
+                    function_type,
+                    return_type_schema,
+                    return_type_name,
+                    return_type_type,
+                    returns_set,
+                    input_arg_types,
+                    input_arg_names
+                ) AS e
+            )
+          ) AS "raw_function_info"
+        FROM
+          hdb_catalog.hdb_function_agg
+        WHERE
+          function_schema = $1 AND function_name = $2
+     |] (sn, fn) False
+
+  case funcData of
+    []                              ->
+      throw400 NotExists $ "no such function exists in postgres : " <>> qf
+    [Identity (Q.AltJ rawFuncInfo)] -> mkFunctionInfo qf rawFuncInfo
+    _                               ->
+      throw400 NotSupported $
+      "function " <> qf <<> " is overloaded. Overloaded functions are not supported"
+
+saveFunctionToCatalog :: QualifiedFunction -> Bool -> Q.TxE QErr ()
+saveFunctionToCatalog (QualifiedObject sn fn) isSystemDefined =
+  Q.unitQE defaultTxErrorHandler [Q.sql|
+         INSERT INTO "hdb_catalog"."hdb_function" VALUES ($1, $2, $3)
+                 |] (sn, fn, isSystemDefined) False
+
+delFunctionFromCatalog :: QualifiedFunction -> Q.TxE QErr ()
+delFunctionFromCatalog (QualifiedObject sn fn) =
+  Q.unitQE defaultTxErrorHandler [Q.sql|
+         DELETE FROM hdb_catalog.hdb_function
+         WHERE function_schema = $1
+           AND function_name = $2
+         |] (sn, fn) False
+
+newtype TrackFunction
+  = TrackFunction
+  { tfName :: QualifiedFunction}
+  deriving (Show, Eq, FromJSON, ToJSON, Lift)
+
+trackFunctionP1
+  :: (CacheRM m, UserInfoM m, QErrM m) => TrackFunction -> m ()
+trackFunctionP1 (TrackFunction qf) = do
+  adminOnly
+  rawSchemaCache <- askSchemaCache
+  when (M.member qf $ scFunctions rawSchemaCache) $
+    throw400 AlreadyTracked $ "function already tracked : " <>> qf
+
+trackFunctionP2Setup :: (QErrM m, CacheRWM m, MonadTx m)
+                     => QualifiedFunction -> m ()
+trackFunctionP2Setup qf = do
+  fi <- withPathK "name" $ liftTx $ getFunctionInfo qf
+  void $ askTabInfo $ fiReturnType fi
```

Comment:
Is this to check if the return table is tracked? If so, it should catch the error and return a more useful error message along these lines: "the return table: tname has to be tracked before the function can be tracked"